### PR TITLE
fix(urd): emits event in constructor

### PR DESCRIPTION
**Pull request by julien-devatom**
[Cantina issue](https://github.com/cantinasec/review-morpho-blue-1/issues/76)

### CODE REVIEW

**Review by Rubilmax — APPROVED** [review]

**Review by QGarchery — COMMENTED** [review]
I think that this would fix #73, because it would become consistent (event are always emitted). What about removing the owner, timelock, root, and ipfsHash arguments to the UrdCreated event though ?

**Review by Jean-Grimal — APPROVED** [review]

**Comment by julien-devatom** [comment]
> I think that this would fix #73, because it would become consistent (event are always emitted). What about removing the owner, timelock, root, and ipfsHash arguments to the UrdCreated event though ?

The problem is that URD constructor events are emitted **before** the factory event. It means that you are not tracking the URD event of the constructor when they are cast. Another solution would be to move the Factory event at the beginning of the function (and we need to precompute the urd address)

**Review by QGarchery — CHANGES_REQUESTED** [review]
> The problem is that URD constructor events are emitted before the factory event

Oh I did not realize that. It's weird to have events that are not used (and also the pending root should actually not emit any event in the constructor). Then I think the easiest solution is to just unfactorize the `_set*` functions in the URD constructor, this is only adding one line of code and no real duplication tbh

**Comment by Rubilmax** [comment]
Why aren't we removing the redundant data from the `UrdCreated` event instead? I don't think they bring much value

**Comment by MerlinEgalite** [comment]
> Why aren't we removing the redundant data from the `UrdCreated` event instead? I don't think they bring much value, because they are only temporary

💯 let's do that

**Comment by julien-devatom** [comment]
> Why aren't we removing the redundant data from the `UrdCreated` event instead? I don't think they bring much value, because they are only temporary

For the same reason explained as before:
The problem is that URD constructor events are emitted before the factory event. 

In fact, when you're indexing from a factory, you start indexing a new contract on `UrdCreated` event. So the indexing mechanism cannot handle the events emitted before. 


However, they are still useful if you are indexing only the URD contract, and that you know in advance the contract address.

**Comment by MerlinEgalite** [comment]
But we can just put events before the construction no?

**Comment by Rubilmax** [comment]
> > Why aren't we removing the redundant data from the `UrdCreated` event instead? I don't think they bring much value, because they are only temporary
> 
> For the same reason explained as before: The problem is that URD constructor events are emitted before the factory event.
> 
> In fact, when you're indexing from a factory, you start indexing a new contract on `UrdCreated` event. So the indexing mechanism cannot handle the events emitted before.
> 
> However, they are still useful if you are indexing only the URD contract, and that you know in advance the contract address.

Alright then I understand that having redundant data serves 2 purposes and that both cases are useful so we should simply acknowledge the issue and do nothing?

**Comment by julien-devatom** [comment]
> Or we can pre-calculate the URD address and simply emit it along the caller in UrdCreated before but I find it ugly


Yeah its weird.. I prefer to keep the 2 events.

**Comment by QGarchery** [comment]
Just doing `owner = newOwner` (and the like) instead of `_setOwner(newOwner)` in the constructor is also a solution, and it seems easy enough

**Comment by Rubilmax** [comment]
> Just doing `owner = newOwner` (and the like) instead of `_setOwner(newOwner)` in the constructor is also a solution, and it seems easy enough

There's a purpose for it too: being able to have access to the initial value of `owner` at construction via `OwnerSet` events emitted from the URD
Otherwise as an integrator you need to also look for the `UrdCreated` event in the factory, which is cumbersome

**Comment by QGarchery** [comment]
> There's a purpose for it too: being able to have access to the initial value of owner at construction via OwnerSet events emitted from the URD

Alright, alright, let's keep it :D

**Review by QGarchery — APPROVED** [review]

**Review by MerlinEgalite — APPROVED** [review]
Just realized that the contract would have inherited from `Ownable` from OZ

**Comment by QGarchery** [comment]
> Just realized that the contract would have inherited from Ownable from OZ

You mean "could have" right ?

**Comment by julien-devatom** [comment]
> Just realized that the contract would have inherited from `Ownable` from OZ

Already discussed in #44

**Comment by MerlinEgalite** [comment]
> > Just realized that the contract would have inherited from `Ownable` from OZ
> 
> Already discussed in #44

Now that there's the version 0.5 of OZ I think we don't have our problem anymore

**Comment by MerlinEgalite** [comment]
Fixed by https://github.com/morpho-org/universal-rewards-distributor/pull/82

File: test/UniversalRewardsDistributorTest.sol [hunk 1]
```diff
@@ -87,33 +87,32 @@ contract UniversalRewardsDistributorTest is Test {
...
-        vm.prank(randomCreator);
+        vm.prank(owner);
         vm.expectEmit(address(urdAddress));
-        emit EventsLib.RootSet(DEFAULT_ROOT, DEFAULT_IPFS_HASH);
+        emit EventsLib.TimelockSet(timelock);
         new UniversalRewardsDistributor{salt: SALT}(
-            randomCreator, DEFAULT_TIMELOCK, DEFAULT_ROOT, DEFAULT_IPFS_HASH
+            owner, timelock, DEFAULT_ROOT, DEFAULT_IPFS_HASH
         );
     }
```
Comment (Rubilmax) [inline test/UniversalRewardsDistributorTest.sol:117]: We could merge these 2 tests into a single one, testing all events
It avoids having to repeat the test case

**Review by Rubilmax — APPROVED** [review]

**Pull request merged**